hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -696,3 +696,141 @@ def valid_tuple_axes(ndim, min_size=0, max_size=None):
+        self.side_strat = st.integers(min_side, max_side)
+        self.min_dims = min_dims
+        self.max_dims = max_dims
+        self.min_side = min_side
+        self.max_side = max_side
+
+    def do_draw(self, data):
+        elements = cu.many(
+            data,
+            min_size=self.min_dims,
+            max_size=self.max_dims,
+            average_size=min(
+                max(self.min_dims * 2, self.min_dims + 5),
+                0.5 * (self.min_dims + self.max_dims),
+            ),
+        )
+        result = []
+        reversed_shape = tuple(self.shape[::-1])
+        while elements.more():
+            if len(result) < len(self.shape):
+                # Shrinks towards original shape
+                if reversed_shape[len(result)] == 1:
+                    if self.min_side <= 1 and not data.draw(st.booleans()):
+                        side = 1
+                    else:
+                        side = data.draw(self.side_strat)
+                elif self.max_side >= reversed_shape[len(result)] and (
+                    not self.min_side <= 1 <= self.max_side or data.draw(st.booleans())
+                ):
+                    side = reversed_shape[len(result)]
+                else:
+                    side = 1
+            else:
+                side = data.draw(self.side_strat)
+            result.append(side)
+        assert self.min_dims <= len(result) <= self.max_dims
+        assert all(self.min_side <= s <= self.max_side for s in result)
+        return tuple(reversed(result))
+
+
+@st.defines_strategy
+def broadcastable_shapes(shape, min_dims=0, max_dims=None, min_side=1, max_side=None):
+    # type: (Sequence[int], int, Optional[int], int, Optional[int]) -> st.SearchStrategy[Tuple[int, ...]]
+    """Return a strategy for generating shapes that are broadcast-compatible
+    with the provided shape.
+
+    Examples from this strategy shrink towards a shape with length ``min_dims``.
+    The size of an aligned dimension shrinks towards being a singleton. The
+    size of an unaligned dimension shrink towards ``min_side``.
+
+    * ``shape`` a tuple of integers
+    * ``min_dims`` The smallest length that the generated shape can possess.
+    * ``max_dims`` The largest length that the generated shape can possess.
+      shape can possess. Cannot exceed 32. The default-value for ``max_dims``
+      is ``2 + max(len(shape), min_dims)``.
+    * ``min_side`` The smallest size that an unaligned dimension can possess.
+    * ``max_side`` The largest size that an unaligned dimension can possess.
+      The default value is 2 + 'size-of-largest-aligned-dimension'.
+
+    The following are some examples drawn from this strategy.
+
+    .. code-block:: pycon
+
+        >>> [broadcastable_shapes(shape=(2, 3)).example() for i in range(5)]
+        [(1, 3), (), (2, 3), (2, 1), (4, 1, 3), (3, )]
+
+    """
+    check_type((tuple, list), shape, "shape")
+    strict_check = max_side is None or max_dims is None
```

Comment:
I introduced `strict_check` because we will because neither `max_side` not `max_dims` will be `None` by the time we get to the broadcastability checks. They will already have had their values set.

Reply:
🤦‍♂️